### ISSUE TITLE
lxc move and "lxc delete error: Failed to run: zfs destroy"

### ISSUE THREAD

**Issue by knutov** [lxc/lxd#9125]
ubuntu 18.04 with hwe kernel (5.4.0-40-generic)
lxd from snap - latest/stable:    4.17
zfs 2.0.4

After `lxc move` container `c6` to the new server (without any errors) I found this container on the old server with STOPPED state.

After `lxc delete` I've got "lxc delete error: Failed to run: zfs destroy" error like in #6815 

After server reboot container was started so it became running on two nodes.

There was no interesting/useful info in any logs and `lxc monitor`, except zfs dataset is busy and there was no processes using it (but zfs dataset presents in some mouninfo):

```
# grep containers/c6 /proc/*/mountinfo
/proc/10679/mountinfo:4246 559 0:219 / /var/snap/lxd/common/shmounts/storage-pools/lxd/containers/c6 rw,noatime shared:288 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/11074/mountinfo:4466 4473 0:219 / /var/snap/lxd/common/shmounts/storage-pools/lxd/containers/c6 rw,noatime shared:288 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/12879/mountinfo:4246 559 0:219 / /var/snap/lxd/common/shmounts/storage-pools/lxd/containers/c6 rw,noatime shared:288 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/13690/mountinfo:4246 559 0:219 / /var/snap/lxd/common/shmounts/storage-pools/lxd/containers/c6 rw,noatime shared:288 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/14007/mountinfo:4732 3873 0:219 / /var/snap/lxd/common/lxd/storage-pools/lxd/containers/c6 rw,noatime shared:540 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/14324/mountinfo:4246 559 0:219 / /var/snap/lxd/common/shmounts/storage-pools/lxd/containers/c6 rw,noatime shared:288 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/16686/mountinfo:4815 4761 0:219 / /var/snap/lxd/common/lxd/storage-pools/lxd/containers/c6 rw,noatime shared:562 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/1803/mountinfo:4246 559 0:219 / /var/snap/lxd/common/shmounts/storage-pools/lxd/containers/c6 rw,noatime shared:288 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/1936/mountinfo:4246 559 0:219 / /var/snap/lxd/common/shmounts/storage-pools/lxd/containers/c6 rw,noatime shared:288 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/21770/mountinfo:4815 4761 0:219 / /var/snap/lxd/common/lxd/storage-pools/lxd/containers/c6 rw,noatime shared:562 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/2326/mountinfo:4732 3873 0:219 / /var/snap/lxd/common/lxd/storage-pools/lxd/containers/c6 rw,noatime shared:540 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/26561/mountinfo:4466 4473 0:219 / /var/snap/lxd/common/shmounts/storage-pools/lxd/containers/c6 rw,noatime shared:288 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/2832/mountinfo:122 3977 0:219 / /var/snap/lxd/common/shmounts/storage-pools/lxd/containers/c6 rw,noatime shared:288 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/30786/mountinfo:4815 4761 0:219 / /var/snap/lxd/common/lxd/storage-pools/lxd/containers/c6 rw,noatime shared:562 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/32537/mountinfo:122 3977 0:219 / /var/snap/lxd/common/shmounts/storage-pools/lxd/containers/c6 rw,noatime shared:288 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/6107/mountinfo:4246 559 0:219 / /var/snap/lxd/common/shmounts/storage-pools/lxd/containers/c6 rw,noatime shared:288 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
/proc/9027/mountinfo:4246 559 0:219 / /var/snap/lxd/common/shmounts/storage-pools/lxd/containers/c6 rw,noatime shared:288 - zfs tank/lxd/containers/c6 rw,xattr,posixacl
root@d1:~# ps aux | grep c6
root     30017  0.0  0.0  14432   968 pts/2    S+   12:10   0:00 grep --color=auto c6
```

` /var/snap/lxd/common/lxd/logs/lxd.log` was not saved before reboot unfortunately.

Suggestions - may be report (at least to the stderr) any errors during `lxc move` as a consequences can be very sad in some cases.

**Comment by stgraber** [comment]
`lxc move` does wait for and report errors from both the source and target servers.

We'd really need the log from the source server to see why the move operation was considered successful in this case.

**Comment by knutov** [comment]
Is there any suitable log I can found after the server was rebooted?

**Comment by stgraber** [comment]
lxd.log is rotated so you can find lxd.log.1 or even .gz compressed versions for older startups.

**Comment by knutov** [comment]
what should I grep from that log?

```
zcat lxd.log.2.gz | grep 'instance=c6'


t=2021-08-18T12:03:15+0300 lvl=dbug msg="MigrateInstance started" args="&{Name:c6 Snapshots:[] MigrationType:{FSType:ZFS Features:[compress]} TrackProgress:true MultiSync:true FinalSync:false Data:<nil> ContentType:}" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:03:30+0300 lvl=dbug msg="GetInstanceUsage started" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:03:30+0300 lvl=dbug msg="GetInstanceUsage finished" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:03:59+0300 lvl=dbug msg="GetInstanceUsage started" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:03:59+0300 lvl=dbug msg="GetInstanceUsage finished" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:04:17+0300 lvl=dbug msg="MigrateInstance finished" args="&{Name:c6 Snapshots:[] MigrationType:{FSType:ZFS Features:[compress]} TrackProgress:true MultiSync:true FinalSync:false Data:map[filesystem:tank/lxd/containers/c6@migration-c755a89d-4780-4957-9b1e-b13557c878dd] ContentType:}" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:04:17+0300 lvl=dbug msg="MigrateInstance started" args="&{Name:c6 Snapshots:[] MigrationType:{FSType:ZFS Features:[compress]} TrackProgress:true MultiSync:true FinalSync:true Data:map[filesystem:tank/lxd/containers/c6@migration-c755a89d-4780-4957-9b1e-b13557c878dd] ContentType:}" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:04:17+0300 lvl=dbug msg="MigrateInstance finished" args="&{Name:c6 Snapshots:[] MigrationType:{FSType:ZFS Features:[compress]} TrackProgress:true MultiSync:true FinalSync:true Data:map[filesystem:tank/lxd/containers/c6@migration-c755a89d-4780-4957-9b1e-b13557c878dd] ContentType:}" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:04:19+0300 lvl=dbug msg="Stop started" instance=c6 instanceType=container project=default stateful=false
t=2021-08-18T12:04:19+0300 lvl=info msg="Stopping container" action=stop created=2021-04-06T05:50:08+0300 ephemeral=false instance=c6 instanceType=container project=default stateful=false used=2021-04-06T05:50:17+0300
t=2021-08-18T12:04:49+0300 lvl=info msg="Stopped container" action=stop created=2021-04-06T05:50:08+0300 ephemeral=false instance=c6 instanceType=container project=default stateful=false used=2021-04-06T05:50:17+0300
t=2021-08-18T12:04:49+0300 lvl=dbug msg="Stop finished" instance=c6 instanceType=container project=default stateful=false
t=2021-08-18T12:04:49+0300 lvl=info msg="Deleting container" created=2021-04-06T05:50:08+0300 ephemeral=false instance=c6 instanceType=container project=default used=2021-04-06T05:50:17+0300
t=2021-08-18T12:04:49+0300 lvl=dbug msg="DeleteInstance started" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:04:49+0300 lvl=dbug msg="Deleting instance volume" driver=zfs instance=c6 pool=lxd project=default volName=c6
t=2021-08-18T12:04:50+0300 lvl=dbug msg="GetInstanceUsage started" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:04:50+0300 lvl=dbug msg="GetInstanceUsage finished" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:04:59+0300 lvl=dbug msg="DeleteInstance finished" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:05:09+0300 lvl=dbug msg="GetInstanceUsage started" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:05:09+0300 lvl=dbug msg="GetInstanceUsage finished" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:06:52+0300 lvl=dbug msg="GetInstanceUsage started" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:06:52+0300 lvl=dbug msg="GetInstanceUsage finished" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:07:00+0300 lvl=info msg="Deleting container" created=2021-04-06T05:50:08+0300 ephemeral=false instance=c6 instanceType=container project=default used=2021-04-06T05:50:17+0300
t=2021-08-18T12:07:00+0300 lvl=dbug msg="DeleteInstance started" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:07:00+0300 lvl=dbug msg="Deleting instance volume" driver=zfs instance=c6 pool=lxd project=default volName=c6
t=2021-08-18T12:07:10+0300 lvl=dbug msg="DeleteInstance finished" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:08:04+0300 lvl=info msg="Deleting container" created=2021-04-06T05:50:08+0300 ephemeral=false instance=c6 instanceType=container project=default used=2021-04-06T05:50:17+0300
t=2021-08-18T12:08:04+0300 lvl=dbug msg="DeleteInstance started" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:08:04+0300 lvl=dbug msg="Deleting instance volume" driver=zfs instance=c6 pool=lxd project=default volName=c6
t=2021-08-18T12:08:14+0300 lvl=dbug msg="DeleteInstance finished" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:29:55+0300 lvl=dbug msg="GetInstanceUsage started" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:29:55+0300 lvl=dbug msg="GetInstanceUsage finished" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:35:52+0300 lvl=dbug msg="GetInstanceUsage started" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:35:52+0300 lvl=dbug msg="GetInstanceUsage finished" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:36:25+0300 lvl=info msg="Deleting container" created=2021-04-06T05:50:08+0300 ephemeral=false instance=c6 instanceType=container project=default used=2021-04-06T05:50:17+0300
t=2021-08-18T12:36:25+0300 lvl=dbug msg="DeleteInstance started" driver=zfs instance=c6 pool=lxd project=default
t=2021-08-18T12:36:25+0300 lvl=dbug msg="Deleting instance volume" driver=zfs instance=c6 pool=lxd project=default volName=c6
t=2021-08-18T12:36:35+0300 lvl=dbug msg="DeleteInstance finished" driver=zfs instance=c6 pool=lxd project=default
```

at the end of log there was multiple trying to do `lxc delete c6`, this container still was not deleted at the end of log.

**Comment by knutov** [comment]
Another container same problem, but there was ther error after `lxc move` this time - should I close this and open another issue about "zfs dataset is busy"?

```
# lxc move ${ct:?} ${dst:?}:${ct:?} -s ${storage:?} --stateless
Error: Failed to delete original instance after copying it: Error deleting storage volume: Failed to run: zfs destroy -r tank/lxd/containers/c2042: cannot destroy snapshot tank/lxd/containers/c2042@migration-a26da0b4-20a6-45dd-968b-089e2f053669: dataset is busy
```

```
grep containers/c2042 /proc/*/mountinfo
```
returns nothing this time

**Comment by stgraber** [comment]
In this case, what's busy is a snapshot, can you check whether you have any `zfs send` or `zfs receive` process running on that system?

**Comment by knutov** [comment]
yes:

```
# ps aux | grep zfs
root     30678  0.0  0.0   9072  4656 ?        S    16:21   0:00 zfs send -c -L -i tank/lxd/containers/c2062@migration-4aa168b5-93c5-46c2-bff0-a355bb7a6ffa tank/lxd/containers/c2062@migration-2d85664e-f9c2-4ca4-a420-1e938679fc37

# strace -p 20674
strace: Process 20674 attached
write(2, "warning: cannot send 'tank/lxd/c"..., 113) = 113
close(3)                                = 0
close(4)                                = 0
close(5)                                = 0
exit_group(1)                           = ?
+++ exited with 1 +++
```

**Comment by stgraber** [comment]
Ok, so that's what would have kept it busy. Now that it's exited, the container should delete just fine.
Any chance you can retry with a clean setup, making sure both sides don't have any leftover `zfs send` or `zfs receive` and attempt the transfer again to see if something gets stuck again?

Our testsuite does hundreds of such migrations in our test runs and other than a few very rare issues with the ZFS tools breaking (I'd guess around once in 50000 or so attempts so far), we've not seen any leftover processes.

**Comment by knutov** [comment]
That is production setup, so it is hard to repeat again due to reboots of containers.

The strange thing - those processes exitted only after I do `strace -p` them. It does not look normal (but this is may be zfs only side).

As for now both servers are freshly booted, the only differences are one with Ubuntu 20.04 and latest hwe kernlel, another with 18.04 and HWE. 
ZFS is zfs-2.1.0-0york2 and LXD is the same 4.17 on both servers.

May I be sure all files/data were trasferred during `lxd move` in this cases?

Is there any way to do `lxc move` like "everything is ok at the end or 'fail to move' and leave container on the old node" mode?

**Comment by stgraber** [comment]
Yeah `strace` unblocking things is very very often an indication of a kernel issue as strace doesn't alter the process' behavior in general.

`lxc move` only deletes the source if the target reports that the migration is complete and didn't hit any errors, which in this case would mean that the datasets were correctly received.

**Comment by stgraber** [comment]
Sounds good. We certainly don't test the DKMS versions of ZFS on old kernels. Our ZFS 2.1 systems currently are on 5.13 I believe, 2.0 comes with 5.11 and you get 0.8 with 5.4 I think (we rely on whatever ZFS is in the Ubuntu kernel as that's got testing, validation and support from the kernel team).

**Comment by stgraber** [comment]
@knutov any update?

**Comment by knutov** [comment]
@stgraber yes, we did upgrade two days ago and now all our servers are ubuntu 20.04 with 5.11 hwe kernel and zfs 2.1.0 from ppa:jonathonf/zfs.

The problem is not reproducible in this setup.

So it looks like it was something about zfs + 5.4 kernel.
